### DISCUSSION TITLE
Custom Provider Import Issues

### DISCUSSION THREAD

**Discussion by kellerdt** [apache/airflow#34895] [Q&A] [answered]
I am running into warnings in several IDE's that imports could not be resolved for a custom airflow providers package that I am developing.  The code itself works when installed as a library on our airflow projects so there isn't anything wrong with the python I think it is a project setup issue with namespaced packages that I was hoping someone could help with.

My library project looks as such:
```
 - airflow
     - providers
         - provider_package_name
             - dags
                 - __init__.py
                 - dag_functions.py 
             - __init__.py
             - utils.py
```

The pyproject.toml contains the following dependency:
apache-airflow = "<version>" 

The import issue is surfaced when I try to import into the "dag_functions.py" from the "utils.py" package like such.
`from airflow.providers.provider_package_name.utils import utility_function`

I could just ignore these errors but it makes development more annoying because I have to manually specify all my dependencies and auto-import functionality doesn't work.

Does anyone have any idea's why my name-spaced packaging isn't resolving in the IDE?  I believe I have the correct `__init__.py` files defined.

**Comment by potiuk** [comment, marked as answer]
You should not keep your custom providers in "airflow.providers" and namespace package. Those are generally foreseen for the "airflow" community providers and there is a custom dynamic import code to make it works well when you import the "airflow" project.  Just use your own package and avoid using namespace packages. Place `__init__.py` in all your packages.
